product_multi_price/models/product_multi_price.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2020 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, api, fields, models, tools
+from odoo.exceptions import ValidationError
+from odoo.addons import decimal_precision as dp
+
+
+class ProductMultiPrice(models.Model):
+    _name = 'product.multi.price'
+
+    name = fields.Many2one(
+        comodel_name='product.multi.price.name',
+        required=True,
+    )
+    product_id = fields.Many2one(
+        comodel_name='product.product',
+        required=True,
+        ondelete='cascade',
+    )
+    price = fields.Float(
+        digits=dp.get_precision('Product Price'),
+    )
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        related='name.company_id',
+        store=True,
+        readonly=True,
+    )
+
+    _sql_constraints = [
+        ('multi_price_uniq', 'unique(name, product_id, company_id)',
+         'A field name cannot be assigned to a product twice for the same'
+         'company'),
+    ]
+
+
+class ProductMultiPriceName(models.Model):
+    _name = 'product.multi.price.name'
+
+    @api.model
+    def _get_company(self):
+        return self._context.get('company_id', self.env.user.company_id.id)
+
+    name = fields.Char(
+        required=True,
+        string='Price Field Name',
+        ondelete='restrict'
+    )
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        default=lambda self: self._get_company()
+    )
+
+    _sql_constraints = [
+        ('multi_price_name_uniq', 'unique(name, company_id)',
+         'Filed Names must be unique per company'),
+    ]
+
+    @api.model
+    @tools.ormcache()
+    def _get_field_names(self):
+        return set([x.name for x in self.search([])])
+
+    @api.constrains('name')
+    def _check_name(self):
+        """The target is to use multi price as if they were virtual fields
+           so we want to constrain the naming to the same rules"""
+        product_fields = list(self.env['product.product']._fields)
```

Comment:
This shouldn't be needed anymore

Reply:
Well, I think is good to have this constraint to avoid:

- setting a price field names with a name already in product/template fields.
- setting a price field name that doesn't comply with field names rules.